### PR TITLE
Remove optional-base plugin

### PR DESCRIPTION
**Pull request by hotchemi**
## Issues
- Resolves https://github.com/hotchemi/PermissionsDispatcher/issues/299, https://github.com/hotchemi/PermissionsDispatcher/issues/304

## Overview

- Remove optional-base plugin and bundle v13 support lib
- Update minSdkVersion to 14
- Remove bundled progurad files
- Update Gradle version to 3.3
  - remove test13 dir and move some test cases to under processor/test/java

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,9 +2,6 @@
 local.properties
 
 build/
-captures/
 
 .idea/
 *.iml
-
-.DS_Store
```
Comment (hotchemi) [inline .gitignore]: just for cleanup

File: .travis.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ branches:
   except:
     - gh-pages
 
-script: ./gradlew check --stacktrace
+script: ./gradlew clean check --stacktrace
```
Comment (hotchemi) [inline .travis.yml]: Just refactor

File: README.md [hunk 3]
```diff
@@ -183,18 +183,6 @@ public class MainActivity extends AppCompatActivity {
 
 ## Note
 
-- PermissionsDispatcher depends on the `support-v4` library by default, in order to be able to use some permission compat classes.
```
Comment (hotchemi) [inline README.md]: I think these lines are no longer needed.

File: library/build.gradle [hunk 4]
```diff
@@ -53,7 +51,7 @@ publish {
     groupId = GROUP_ID
     artifactId = ARTIFACT_ID_LIBRARY
     version = VERSION
-    description = DESCRIPTION
+    desc = DESCRIPTION
```
Comment (hotchemi) [inline library/build.gradle]: description became deprecated.

File: processor/src/main/kotlin/permissions/dispatcher/processor/util/Helpers.kt [hunk 5]
```diff
@@ -25,7 +25,7 @@ fun permissionRequestTypeName(e: ExecutableElement): String = "${e.simpleString(
 fun <A : Annotation> findMatchingMethodForNeeds(needsElement: ExecutableElement, otherElements: List<ExecutableElement>, annotationType: Class<A>): ExecutableElement? {
     val value: List<String> = needsElement.getAnnotation(NeedsPermission::class.java).permissionValue()
     return otherElements.firstOrNull {
-        it.getAnnotation(annotationType).permissionValue().equals(value)
+        it.getAnnotation(annotationType).permissionValue() == value
```
Comment (hotchemi) [inline processor/src/main/kotlin/permissions/dispatcher/processor/util/Helpers.kt]: Just refactor

File: processor/src/main/kotlin/permissions/dispatcher/processor/util/Validators.kt [hunk 6]
```diff
@@ -38,7 +38,7 @@ fun <A : Annotation> checkDuplicatedValue(items: List<ExecutableElement>, annota
         val permissionValue = it.getAnnotation(annotationClass).permissionValue()
         Collections.sort(permissionValue)
         allItems.forEach { oldItem ->
-            if (oldItem.equals(permissionValue)) {
+            if (oldItem == permissionValue) {
```
Comment (hotchemi) [inline processor/src/main/kotlin/permissions/dispatcher/processor/util/Validators.kt]: Just refactor

File: processor/src/testV13/java/permissions/dispatcher/processor/V13ProcessorTestSuite.java [hunk 7]
```diff
@@ -1,104 +0,0 @@
-package permissions.dispatcher.processor;
-
-import org.junit.Test;
-import permissions.dispatcher.processor.base.TestSuite;
-import permissions.dispatcher.processor.data.V13Source;
-
-public class V13ProcessorTestSuite extends TestSuite {
```
Comment (hotchemi) [inline processor/src/testV13/java/permissions/dispatcher/processor/V13ProcessorTestSuite.java]: This class has not been executed for a long time...OMG 🙀

**Comment by aurae** [comment]
In general, I approve the changes you made (raising minSdkVersion, Kotlin clean-up). However, I see that you removed the safeguards around `FragmentCompat`. As stated [here](https://github.com/hotchemi/PermissionsDispatcher/issues/304#issuecomment-298185071), this will cause an issue, because we can't really make usage of `support-v13` optional if the pre-requisite checks and `V13Access` are removed.

I had to pull the branch locally and test this out, but I was able to confirm this somewhat. Here's the example:

```groovy
// Abbreviated; I don't want native fragments, so I exclude the dependency
compile ("permissions-dispatcher:2.4.0") {
    exclude module: "support-v13"
}
annotationProcessor "..."
```

```java
@RuntimePermissions
class NativeFragment extends android.app.Fragment {
    // In this example, I'm an incosistent user that still annotates a native fragment
}
```

At runtime, the first call to `NativeFragment#xxxWithCheck()` will crash with a `ClassNotFoundException` here, since `FragmentCompat` is missing. Before, the `V13Access` class and `NativeFragmentProcessorUnit` made sure to check if it's there before trying anything.

Ideally, the annotation processor is aware of this inconsistency (but that probably won't work, since we can't rely on `FragmentCompat` on the classpath, which is the reason that issues like #279 even exist). Otherwise, maybe we can use Lint here! A Lint Detector should be able to discover the inconsistency, where I cannot annotate a native Fragment with `@RuntimePermissions`, if I don't have `FragmentCompat` on the classpath!

As an alternative, we can just remove the whole "optional dependency" thing and always use support-v13. It's only like 300 methods anyway, and ProGuard will take care of it if users don't use it. ¯\_(ツ)_/¯

File: library/build.gradle (hunk 4 above)
Comment (aurae) [inline library/build.gradle]: Good to know! I have to check that in my other libraries :grin:

**Comment by hotchemi** [comment]
@aurae Ah thx for your good catch!
So how about just keep `checkPrerequisites`? With that we'll be able to find the problem in a compile time. Adding lint check would be fine as well though!
https://github.com/hotchemi/PermissionsDispatcher/pull/306/files#diff-de57b3433a39c68273e05dc9453aaf58L18

**Comment by aurae** [comment]
I'm afraid that we're going to get into a situation similar to #279 with `checkPrerequisites`... the annotation processor doesn't seem to share the entire classpath reliably with the application anymore. Therefore, we would not be able to detect the missing `FragmentCompat`, I think. Maybe a Lint check is the way to go (or we just force support-v13, that's definitely the easier option)...

**Comment by hotchemi** [comment]
Okay, so let me merge this PR and create lint later! I'll address https://github.com/hotchemi/PermissionsDispatcher/issues/305 at first!

**Pull request merged**